src/thermo/RedlichKwongMFTP.cpp

```diff
@@ -566,18 +578,85 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
                                "Unknown thermo model : " + model);
         }
 
-        // Go get all of the coefficients and factors in the
-        // activityCoefficients XML block
+        // Go get all of the coefficients and factors in the activityCoefficients XML block
         if (thermoNode.hasChild("activityCoefficients")) {
             XML_Node& acNode = thermoNode.child("activityCoefficients");
 
-            // Loop through the children getting multiple instances of
-            // parameters
+            // Count the number of species with parameters provided in the input file:
+            size_t nParams = 0;
+
+            // Loop through the children and read out fluid parameters.  Process all the
+            // pureFluidParameters, first:
             for (size_t i = 0; i < acNode.nChildren(); i++) {
                 XML_Node& xmlACChild = acNode.child(i);
                 if (ba::iequals(xmlACChild.name(), "purefluidparameters")) {
                     readXMLPureFluid(xmlACChild);
-                } else if (ba::iequals(xmlACChild.name(), "crossfluidparameters")) {
+                    nParams += 1;
+                }
+            }
+
+            // If any species exist which have undefined pureFluidParameters, search the database
+            //     in 'critProperties.xml' to find critical temperature and pressure and calculate
+            //     a and b based on that.
+            if (m_kk > nParams) // activity coefficients of all species are not listed in CTI file. Look into the database.
+            {
+                // Get the array of species names:
+                XML_Node& sparrays = phaseNode.child("speciesArray");
+                std::vector<std::string> spnames;
+                getStringArray(sparrays, spnames);
+
+                // Loop through all species in the CTI file
+                size_t iSpecies = 0;
+
+                for (size_t i = 0; i < m_kk; i++)
+                {
+                    string iName = ba::to_lower_copy(spnames[i]);
+
+                    // Get the index of the species
```

Comment:
Hey, finally working on implementing these, and for some reason this one change does not work.  The lookup function fails when I make this change.

Separate (but possibly related) question: do I actually need to do this circular type loop to then retrieve the species index?  if I'm looping over `k = 0; k < m_kk`, is k going to be equal to the iSpecies that I calculate on line 616?

Reply:
Right, you don't need to call `speciesIndex` to get the index, since you are already just looping over the species (so the `iSpecies` variable should be eliminated). What I am suggesting is that you also shouldn't need to touch the `phaseNode` to get at the species names -- `spnames[i]` should be the same as just calling `speciesName(i)`.